Nested Injectors: Each injector responsible for its instances

davidsdearaujo opened the pull request:
## The problem
I think each module should take care of its binds AND instances.

Let's suppose you create nested injectors: `parentInjector` and `innerInjector`. In this example `innerInjector` have a 
**lazy singleton** bind of `Class1`.

If you ask for `Class1` for `innerInjector` or `parentInjector`, they should return the same instance, right?

Well, it's not happening like that. When the injectors are nested, the main injector make a copy of all the inner injectors to it's own binds, so when you ask to `parentInjector` and `innerInjector` the same bind, they return different instances.

## Expected Behavior
The expected behavior was that the `innerInjector` would be able to create the instances for its binds and `parentInjector` would just ask it to do so.

## What does this PR do
This PR makes each injector take care of it's own **binds** and **instances**. 
When there are nested injectors, the injector find in its own binds at first and if the bind is not found then it ask to the inner injectors.

davidsdearaujo committed:
Nested Injectors
Each injector responsible for its instances

jacobaraujo7 commented on lib/src/auto_injector_base.dart:
```diff
@@ -185,6 +185,7 @@ abstract class AutoInjector extends Injector {
 
 class _AutoInjector extends AutoInjector {
   final _binds = <Bind>[];
+  final _injectorsList = <_AutoInjector>[];
```
Poderia ser um map pra acelerar a busca em que a key seria a tag do AutoInjector

davidsdearaujo replied:
It's not useful anymore because now we are using the **LayersGraph** to find binds and injectors using Graph with [BFS (Breadth First Search)](https://www.geeksforgeeks.org/breadth-first-search-or-bfs-for-a-graph/).

This list only deserves to help **LayersGraph** fill the `adjacencyList`  when `LayersGraph.initialize()` is called.

jacobaraujo7 commented on lib/src/auto_injector_base.dart:
```diff
@@ -289,16 +290,7 @@ class _AutoInjector extends AutoInjector {
-    }
-
-    _paramTransforms.addAll(injector._paramTransforms);
+    _injectorsList.add(injector);
```
atribuir a key do map

davidsdearaujo replied:
It's not useful anymore because now we are using the **LayersGraph** to find binds and injectors using Graph with [BFS (Breadth First Search)](https://www.geeksforgeeks.org/breadth-first-search-or-bfs-for-a-graph/).

This list only deserves to help **LayersGraph** fill the `adjacencyList`  when `LayersGraph.initialize()` is called.

jacobaraujo7 reviewed (changes requested):
Muito bom o seu PR. 
Vou solicitar que você adicione uma forma de remover um AutoInjector interno.
Também é necessário adicionar o get correto no getNotifier().
Also é necessário modificar o replaceInstance() para mudar os binds dos innersInjectors.

davidsdearaujo committed:
Fix bind search by layers
 - Use Graph with BFS to get binds;
 - Fix unique injector tag (using uuid);
 - _AutoInjector: Changed to AutoInjectorImpl
 - AutoInjectorImpl: Make internal variables to be able to test

jacobaraujo7 commented on lib/src/auto_injector_base.dart:
```diff
@@ -183,15 +188,22 @@ abstract class AutoInjector extends Injector {
-  final _binds = <Bind>[];
-  final _injectorsList = <_AutoInjector>[];
-  var _commited = false;
+class AutoInjectorImpl extends AutoInjector {
```
Não me oponhho a deixar a classe publica, desde que seja escondida (hide) no arquivo `lib/auto_injector.dart`

davidsdearaujo replied:
Alright, I'll hide it!
I changed it to public class to be able to test it.

davidsdearaujo replied:
Fixed in this commit: [Make "AutoInjectorImpl" invisible on "package:auto_injector/auto_injector.dart"](https://github.com/davidsdearaujo/auto_injector/commit/3758e9e348faf7db6d12d6fb423cd9e3384d0c95)

davidsdearaujo committed:
Make "AutoInjectorImpl" invisible on
"package:auto_injector/auto_injector.dart"

davidsdearaujo committed:
Binds: Remove tag

davidsdearaujo committed:
Binds: Remove tag

davidsdearaujo committed:
Use LayersGraph to remove bind

davidsdearaujo committed:
Use LayersGraph to validate if a bind exists

davidsdearaujo committed:
Created dispose methods
"dispose()", "disposeRecursive()" and "disposeInjectorByTag()"

davidsdearaujo committed:
Remove injector automatically from parents layersGraph after dispose

davidsdearaujo committed:
Use graph on AutoInjector.isInstantiateSingleton

davidsdearaujo committed:
Replace instance: fix error message

davidsdearaujo committed:
Merge branch 'main' of https://github.com/flutterando/auto_injector

davidsdearaujo committed:
Layers Graph: Make "removeWhenDispose()" private

davidsdearaujo committed:
Merge branch 'main' of https://github.com/flutterando/auto_injector

davidsdearaujo committed:
LayersGraph: Added HashMap

davidsdearaujo committed:
LayersGraph: Use LinkedHashMap
